mantidproject/mantid
# Fix Double Click of Ragged workspaces

**Description of work.**
This PR fixes a bug where an error message appears after double clicking a ragged workspace in the ADS.

The problem was caused by the `blocksize` method (which is not compatible with ragged workspaces). The ongoing issues with ragged workspaces has shown there are several missing methods in the MatrixWorkspace API. @martyngigg has suggested adding:

- `getNumberBins(std::size_t index)` can now be used to find the number of bins for a given histogram index.
- `getMaxNumberBins()` will get the maximum number of bins in a workspace (useful for ragged workspaces)!

These have been added in this PR.

**To test:**
1. Run this code to create a ragged workspace:
```
ws = CreateWorkspace([0,1,2,3,4,2, 3,4,5,6],[0,1,2,3,4,3,4,5,6,7], NSpec=2)
min = 1
max = 6
for j in range(0,2):
    x =[]
    y =[]
    data = ws.readY(j)
    k=0
    for v in ws.readX(j):
        if v >= min and v<=max:
            x.append(v)
            y.append(data[k])
        k+=1
    tmp = CreateWorkspace(x,y)
    ConjoinWorkspaces(ws, tmp, CheckOverlapping=False)
```
2. Double click the ragged workspace in the ADS. The plot dialog should appear, and no error in the logger.

Fixes #30071

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
